tools/shared.py

```diff
@@ -96,6 +99,101 @@ def run_process(cmd, check=True, input=None, *args, **kw):
+  # By default, avoid using Python multiprocessing library due to a large amount of bugs it has on Windows (#8013, #718, #13785, etc.)
+  # Use EM_PYTHON_MULTIPROCESSING=1 environment variable to enable it. It can be faster, but may not work on Windows.
+  if int(os.getenv('EM_PYTHON_MULTIPROCESSING', '0')):
+    import multiprocessing
+    global multiprocessing_pool
+    if not multiprocessing_pool:
+      multiprocessing_pool = multiprocessing.Pool(processes=get_num_cores())
+    return multiprocessing_pool.map(mp_run_process, [(cmd, env, route_stdout_to_temp_files_suffix, pipe_stdout, check, cwd) for cmd in commands], chunksize=1)
+
+  std_outs = []
+
+  if route_stdout_to_temp_files_suffix and pipe_stdout:
+    raise Exception('Cannot simultaneously pipe stdout to file and a string! Choose one or the other.')
+
+  # TODO: Experiment with registering a signal handler here to see if that helps with Ctrl-C locking up the command prompt
+  # when multiple child processes have been spawned.
+  # import signal
+  # def signal_handler(sig, frame):
+  #   sys.exit(1)
+  # signal.signal(signal.SIGINT, signal_handler)
+
+  with ToolchainProfiler.profile_block('run_multiple_processes'):
+    processes = []
+    num_parallel_processes = get_num_cores()
+    temp_files = configuration.get_temp_files()
+    i = 0
+    num_completed = 0
+
+    while num_completed < len(commands):
+      if i < len(commands) and len(processes) < num_parallel_processes:
+        # Not enough parallel processes running, spawn a new one.
+        std_out = temp_files.get(route_stdout_to_temp_files_suffix) if route_stdout_to_temp_files_suffix else (subprocess.PIPE if pipe_stdout else None)
+        if DEBUG:
+          logger.debug('Running subprocess %d/%d: %s' % (i + 1, len(commands), ' '.join(commands[i])))
+        processes += [(i, subprocess.Popen(commands[i], stdout=std_out, stderr=subprocess.PIPE if pipe_stdout else None, env=env, cwd=cwd))]
+        if route_stdout_to_temp_files_suffix:
+          std_outs += [(i, std_out.name)]
+        i += 1
+      else:
+        # Not spawning a new process (Too many commands running in parallel, or no commands left): find if a process has finished.
+        def get_finished_process():
+          while True:
+            j = 0
+            while j < len(processes):
+              if processes[j][1].poll() is not None:
+                out, err = processes[j][1].communicate()
+                return (j, out.decode('UTF-8') if out else '', err.decode('UTF-8') if err else '')
+              j += 1
+            # All processes still running; wait a short while for the first (oldest) process to finish,
+            # then look again if any process has completed.
+            try:
+              out, err = processes[0][1].communicate(0.2)
+              return (0, out.decode('UTF-8') if out else '', err.decode('UTF-8') if err else '')
+            except subprocess.TimeoutExpired:
+              pass
```

Comment:
how about not always checking process 0 (oldest) in the last part here? instead, since we have a timeout, we can have a counter `k` that replaces that 0, and cycles around. Then if the oldest process is very very slow we won't get stuck there. (The problem is probably very rare, but worth avoiding i think.)

also, I think this can be simplified to avoid the duplication of 168-169 and 174-175. instead of the second `communicate(0.2)` we can call `wait(0.2)`. if the wait succeeds we can set `j` and go back to the top of the loop, where the polling will find it. but not sure if that's simpler overall.

Reply:
> Then if the oldest process is very very slow we won't get stuck there.

I don't think that is the right call. If we have an array of processes, and at any time make an assumption that the oldest process should finish first - then after having checked that process, and having found that it has not finished, and no other process has finished either, none of that information will change our assumption that the oldest process should finish first again. I.e. it is optimal to wait for the first process again after we have observed that none of the processes in the array have finished.

Reply:
>  instead of the second `communicate(0.2)` we can call `wait(0.2)`

I had it like that first, but it would deadlock harfbuzz build. Python docs state that [`wait` is broken](https://docs.python.org/3/library/subprocess.html):

```
Note

This will deadlock when using stdout=PIPE or stderr=PIPE and the child process generates enough output to a pipe such that it blocks waiting for the OS pipe buffer to accept more data. Use Popen.communicate() when using pipes to avoid that.
```
Why `python` actually then carries that function around instead of fixing their code is beyond me. However it does deadlock in practice, even if the wait timeout is tiny like 0.2 seconds.

Reply:
Sounds good on both points.

For the first, I missed that it goes back up anyhow in the loop, so yes, this seems optimal (or as optimal as can be without select).